rust-guest/src/context/context.rs

```diff
@@ -0,0 +1,102 @@
+#![allow(dead_code)]
+#![allow(unused_imports)]
+#![allow(unused_variables)]
+
+use lazy_static::lazy_static;
+use std::sync::Mutex;
+
+use std::io::Cursor;
+use std::mem;
+use std::collections::HashMap;
+use super::generated::{Encode, Decode, Context, Request, Response};
+
+lazy_static! {
+    pub static ref PTR_LEN: Mutex<(u32, u32)> = Mutex::new((0, 0));
+}
+
+pub trait RunContext {
+    fn new() -> Self;
+    fn from_read_buffer(self, read_buff_global: &mut Cursor<&mut Vec<u8>>) -> Self;
+    fn to_write_buffer(self) -> (u32, u32);
+    fn next(self) -> Self;
+    fn request(&mut self) -> &mut Request;
+    fn response(&mut self) -> &mut Response;
+    fn default() -> Self;
+}
+
+impl RunContext for Context {
+    fn new()  -> Context {
+          Self::default()
+    }
+
+    fn from_read_buffer(self, read_buff_global: &mut Cursor<&mut Vec<u8>>) -> Self {
+          Decode::decode(read_buff_global).unwrap().unwrap()
+    }
+
+    fn to_write_buffer(self) -> (u32, u32) {
+        let mut cursor = Cursor::new(Vec::new());
+        let _ = Encode::encode(self, &mut cursor);
+
+        let mut vec = cursor.into_inner();
+        vec.shrink_to_fit();
+        let ptr = vec.as_ptr() as u32;
+        let len = vec.len() as u32;
+        mem::forget(vec);  // prevents deallocation in Rust
```

Comment:
calling forget here is fine, but if we don't manually free the memory before we reallocate, we're just going to be leaking memory

Reply:
Yes, definitely a problem. If we free at all before the run() function finishes we won't be able to read the ptr/len on the host side, so wanted to discuss this because we may need to dealloc from the host for rust-specific Wasm modules, which is obviously not a good coupling. Am looking into a better solution for this currently.

fwiw, if this is the last in the chain or by itself, the whole thing will be cleanup by the Wasm runtime, but obviously that doesn't help us if next() is called at all here.

Reply:
This is now mitigated [here](https://github.com/loopholelabs/scale/pull/16/commits/4d5972112389a7e6b59b7a3d211c7373b767d08d#diff-bb6958694f104af4846001b3211bbe9e26dee4e60915447dec4b65a979db5b54R60) which also allows the resize function to be able start wit this buffer with each subsequent module invocation instead of reallocating completely.